Scripts/Enemies/Skeleton/States/attack.gd

```
@@ -2,26 +2,30 @@ class_name SkeletonAttack
 extends State
 
 @export var attack_damage: float = 25.0
-@export var attack_range: float = 40.0
+@export var attack_range: float = 30.0
 @export var attack_cooldown: float = 1.5
 @export var skeleton: CharacterBody2D
 
 var player: CharacterBody2D
 var sprite: AnimatedSprite2D
 var can_attack: bool = true
 var attack_timer: float = 0.0
+var has_dealt_damage: bool = false
+var animation_finished: bool = false
+var attack_duration: float = 0.0
 
 func Enter():
 	
 	if skeleton:
 		sprite = skeleton.get_node("AnimatedSprite2D")
+		if not sprite.frame_changed.is_connected(_on_sprite_frame_changed):
+			sprite.frame_changed.connect(_on_sprite_frame_changed)
 	
 	player = get_tree().get_first_node_in_group("Player")
 	can_attack = true
 	attack_timer = 0.0
-	
-	if sprite:
-		sprite.play("attack")
+	has_dealt_damage = false
+	animation_finished = true
 
 func Update(delta: float):
 	if not skeleton or not is_instance_valid(skeleton):
```

Same as in 2 comments above, is there's any scenario where skeleton is falsy or invalid, then we don't want the state to fail silently, it'll be hell to find out what happened, and the only scenarios where this may happen are if we manually do something wrong, so we want to have a crash rather than a silent fail